services/wallet/history/service.go

```diff
@@ -306,242 +329,207 @@ func tokenToValue(tokenCount *big.Int, mainDenominationValue float32, weisInOneM
-				balance.Add(balance, (*big.Int)(chainBalance.Balance))
-			}
+	for _, address := range addresses {
+		s.triggerEvent(EventBalanceHistoryUpdateStarted, address, "")
 
-			// if last stride, the timestamp might be in the future
-			if strideEnd > time.Now().UTC().Unix() {
-				strideEnd = time.Now().UTC().Unix()
+		for _, network := range networks {
+			if network.IsTest != areTestNetworksEnabled {
+				continue
 			}
 
-			res = append(res, &DataPoint{
-				Timestamp:   uint64(strideEnd),
-				Balance:     (*hexutil.Big)(balance),
-				BlockNumber: (*hexutil.Big)(getBlockID(chainMaxBalance)),
-			})
-		}
+			entries, err := s.balance.db.GetEntriesWithoutBalances(network.ChainID, common.Address(address))
+			if err != nil {
+				log.Error("Error getting blocks without balances", "chainID", network.ChainID, "address", address.String(), "err", err)
+				return err
+			}
 
-		if allPastEnd(data, pos) {
-			return res, nil
-		}
+			log.Debug("Blocks without balances", "chainID", network.ChainID, "address", address.String(), "entries", entries)
 
-		strideStart = strideEnd
-	}
-}
+			client, err := s.rpcClient.EthClient(network.ChainID)
+			if err != nil {
+				log.Error("Error getting client", "chainID", network.ChainID, "address", address.String(), "err", err)
+				return err
+			}
 
-func getBlockID(chainBalance map[chainIdentity]*DataPoint) *big.Int {
-	var res *big.Int
-	for _, balance := range chainBalance {
-		if res == nil {
-			res = new(big.Int).Set(balance.BlockNumber.ToInt())
-		} else if res.Cmp(balance.BlockNumber.ToInt()) != 0 {
-			return nil
+			err = s.addEntriesToDB(ctx, client, network, address, entries)
+			if err != nil {
+				return err
+			}
 		}
+		s.triggerEvent(EventBalanceHistoryUpdateFinished, address, "")
 	}
 
-	return res
+	log.Debug("updateBalanceHistory finished")
+	return nil
 }
 
-type timeIdentity struct {
-	chain chainIdentity
-	index int
-}
+func (s *Service) addEntriesToDB(ctx context.Context, client chain.ClientInterface, network *params.Network, address statustypes.Address, entries []*entry) (err error) {
+	for _, entry := range entries {
+		var balance *big.Int
+		// tokenAddess is zero for native currency
+		if (entry.tokenAddress == common.Address{}) {
+			// Check in cache
+			balance = s.balanceCache.GetBalance(common.Address(address), network.ChainID, entry.block)
+			log.Debug("Balance from cache", "chainID", network.ChainID, "address", address.String(), "block", entry.block, "balance", balance)
 
-func (i timeIdentity) dataPoint(data map[chainIdentity][]*DataPoint) *DataPoint {
-	return data[i.chain][i.index]
-}
+			if balance == nil {
+				balance, err = client.BalanceAt(ctx, common.Address(address), entry.block)
+				if balance == nil {
+					log.Error("Error getting balance", "chainID", network.ChainID, "address", address.String(), "err", err, "unwrapped", errors.Unwrap(err))
+					return err
+				}
+				time.Sleep(50 * time.Millisecond) // TODO Remove this sleep after fixing exceeding rate limit
```

Comment:
It works for now, and it would be great if we have a task to track this workaround.
Indeed, at first glance, requests/second quota should not be part of the responsibilities of RPC clients and should be offloaded to a throttler who knows the API requirements for each service and handles this transparently.

Reply:
The one you mentioned in your comment.